### PR TITLE
 Add more constructor to C api for shape 

### PR DESCRIPTION
**Pull request by pfultz2**

### CODE REVIEW

File: src/api/api.cpp [hunk 1]
```diff
@@ -611,16 +629,16 @@ extern "C" migraphx_status migraphx_onnx_options_create(migraphx_onnx_options_t*
...
+    migraphx_onnx_options_t onnx_options, const char* name, size_t* dims, size_t dims_size)
 {
     return migraphx::try_([&] {
         if(onnx_options == nullptr)
             MIGRAPHX_THROW(migraphx_status_bad_param, "Bad parameter onnx_options: Null pointer");
-        migraphx::set_input_parameter_shape((onnx_options->object), (name), (dims), (dim_num));
+        if(dims == nullptr)
+            MIGRAPHX_THROW(migraphx_status_bad_param, "Bad parameter dims: Null pointer");
+        migraphx::set_input_parameter_shape(
+            (onnx_options->object), (name), (std::vector<size_t>(dims, dims + dims_size)));
```
Comment (scxiao) [inline src/api/api.cpp:641]: This change revert to the code in my previous PR parse_onnx_using_explicit_input_shape. The problem here is that if dim is nullptr, an exception is thrown.

But in the function onnx_options::set_input_parameter_shape(name, dims.data(), dim.size()) in migraphx.hpp, if dim is empty (used to represent a scalar shape in both numpy and ort), we will throw an exception.
That is the reason I made some changes in my PR python_scalar_input_shape.

File: src/api/include/migraphx/migraphx.h [hunk 2]
```diff
@@ -77,6 +77,13 @@ migraphx_status migraphx_shape_create(migraphx_shape_t* shape,
                                       size_t* lengths,
                                       size_t lengths_size);
 
+migraphx_status migraphx_shape_create_with_strides(migraphx_shape_t* shape,
+                                                   migraphx_shape_datatype_t type,
+                                                   size_t* lengths,
+                                                   size_t lengths_size,
+                                                   size_t* strides,
+                                                   size_t strides_size);
+
```
Comment (scxiao) [inline src/api/include/migraphx/migraphx.h:86]: This function still cannot handle the case of a scalar shape for the function set_input_parameter_shape(name, dims), where dims is a empty vector

File: src/api/migraphx.py [hunk 3]
```diff
@@ -176,9 +181,7 @@ def onnx_options(h):
     h.constructor('create')
     h.method(
         'set_input_parameter_shape',
-        api.params(name='const char*',
-                   dims='const size_t *',
```
Comment (scxiao) [inline src/api/migraphx.py:180]: Same as above, the reason we change to this API is to process the case of dims.datat() is nullptr.

File: test/api/test_cpu.cpp [hunk 4]
```diff
@@ -64,21 +64,22 @@ TEST_CASE(zero_parameter)
...
-    CHECK(bool{s1 == s1_orig});
+    auto s = migraphx::shape(migraphx_shape_float_type);
+    EXPECT(s.lengths().size() == 1);
+    EXPECT(s.strides().size() == 1);
+    EXPECT(s.lengths().front() == 1);
+    EXPECT(s.strides().front() == 0);
+}
 
-    migraphx::onnx_options option;
-    option.set_input_parameter_shape("1", {});
```
Comment (scxiao) [inline test/api/test_cpu.cpp:76]: This line will throw an exception and this is what we meet in ORT

File: src/api/api.cpp (hunk 1 above)
Comment (pfultz2) [reply src/api/api.cpp:641]: Sorry, I am not sure why this was reverted, maybe a merge mistake. I will try to fix it.

File: test/api/test_cpu.cpp (hunk 4 above)
Comment (pfultz2) [reply test/api/test_cpu.cpp]: I added this test back in and I fixed the null pointer error.

**Pull request merged**